allows info_pattern to match empty values. Fixes #234

jsmedmar opened the pull request:
Regards,

On Issue #234 I described an error when reading INFO tags. This pull request allows the `info_pattern` regex to match empty values. It now works for my vcf files and perhaps you may want to marge the changes. 

Let me know if you have any questions.

martijnvermaat commented:
Hi @jamescasbon, thanks for the PR! :+1: 

Do you know how the VCF file was created? If not, where did you get it from? If more users can expect this to happen I would be happy to merge this.

Could you include a test case? Let me know if you need help with that.

martijnvermaat commented on vcf/parser.py:
```diff
@@ -83,7 +83,7 @@ def __init__(self):
         self.info_pattern = re.compile(r'''\#\#INFO=<
             ID=(?P<id>[^,]+),\s*
-            Number=(?P<number>-?\d+|\.|[AGR]),\s*
+            Number=(?P<number>(?:(-?\d+|\.|[AGR]))?),\s*
```
I think it would be enough to just append a `?` to the matching group, like this:

    Number=(?P<number>-?\d+|\.|[AGR])?,\s*

This should make the match group value `None` for empty values, obsoleting the extra check in `vcf_field_count` below.

jsmedmar commented:
I obtained this VCF file using the Caveman pipeline from Wellcome Sanger's institute. 

As you suggested, I changed this ``Number=(?P<number>(?:(-?\d+|\.|[AGR]))?),\s*`` for ``Number=(?P<number>-?\d+|\.|[AGR])?,\s*`` and remove the unnecessary check.

I also added the following test under `vcf/test/test_vcf.py`:

```
class TestIssue234(unittest.TestCase):

    def test_read_info_handles_empty_values_for_number(self):
        """``_vcf_metadata_parser`` shouldnt break if Number tag is None."""
        parser = vcf.parser._vcf_metadata_parser()
        num_str = '##INFO=<ID=CA,Number=,Type=Flag,Description="Position '
        num_str += 'could not be annotated to a coding region of a transcript '
        num_str += 'using the supplied bed file">'
        try:
            parser.read_info(num_str)
        except SyntaxError:
            msg = "vcf.parser._vcf_metadata_parser shouldn't raise SyntaxError"
            msg += " if Number tag is empty."
            self.fail(msg)
```

However, when I run `tox` my test is not being executed. Do you have any idea of why this is happening?

martijnvermaat commented:
Are the other tests executed? Then you should probably add your test to the suite at the very bottom of `test_vcf.py`. I make this mistake all the time and I think the standard unittest setup violates DRY in many ways...

If you could also `assert` that the `num` field in the returned `Info` instance is `None`, this looks good to me.

jsmedmar commented:
Yes I'm sorry, I missed the fact that I needed to register my test case...

jsmedmar commented:
Ok, test is passing I think we are good to go. This is the pipeline I'm talking about: https://github.com/cancerit/CaVEMan, we are currently using it at Memorial Sloan Kettering in NYC.

martijnvermaat commented:
Great @jsmedmar! (sorry to address the wrong person in my first comment)

Had a look in [this repository for CaVEMan](https://github.com/cancerit/CaVEMan) and I couldn't find any part of the code that would produce this (grepped their entire Git history for the example you provided). Is there another Caveman pipeline, or where you using a version that's not in the repository?

I'm just trying to get a better understanding on where this is coming from and perhaps see if we should generalize this to the other attributes and/or `FORMAT` definitions.

jsmedmar commented:
We use exactly these repos:

- Caveman Wrapper [Version 1.7.5]: https://github.com/cancerit/cgpCaVEManWrapper
- Caveman Post Processing [Version 1.5.2]: https://github.com/cancerit/cgpCaVEManPostProcessing

I think I also had this problem with CGP's Pindel. Let me test it with PyVCF master's repo and I'll let you know.

jsmedmar commented:
I just test my Pindel vcf files (a pipeline also from Sanger's CGP) and they worked properly with PyVCF pip distribution. Therefore I'm only getting this error with the Caveman output.

 I think this change is not likely to bring problems to other users. But the concern you raise is interesting, maybe there is no harm in accepting empty values for other tags as well.

martijnvermaat commented:
My Perl is pretty rusty, but my feeling is that the issue was fixed here:

https://github.com/cancerit/cgpCaVEManPostProcessing/commit/e8c11d17b5ebe873213723b66f90927218ff6cbb#diff-709fe86b6172dba36167dc379cd90db7R94

That would be consistent with you using cgpCaVEManPostProcessing 1.5.4 or older.

Since the syntax is clearly in violation with the [VCF spec](https://samtools.github.io/hts-specs/VCFv4.2.pdf) and it doesn't seem to occur in any other locations, I don't think we should now generalize this to other tags.

But as you said, I think your fix is unlikely to do any harm, so I'll merge it.

Thanks and enjoy your NYC evening sun, I'll be visiting the city next month :smile:

jsmedmar commented:
NYC's evenings are beautiful these days! You should stop by MSK's Faculty club to have a beer either Friday or Thursday! (medinaj@mskcc.org)
